### PR TITLE
add TrOneIsOpaque option to MTLLoader

### LINKED ISSUE #12861: add TrOneIsOpaque option to MTLLoader
##### Description of the problem 

Today there are two major definitions of keyword `Tr` in mtl. And both of them are applied in different big 3D softwares. So I think it a proper way to add a new option `equalizeTransparency` in `materialOptions` of `MTLLoader`.

##### Details of this issue
According to the original spec http://paulbourke.net/dataformats/mtl/ , there's no Tr field. The Tr field is added into mtl format later than other field, and up to now, there are not any 'standard' spec defined it.

The most widely used definition might be set Tr = 1 - d. The value of 1.0 of d or the 0 of Tr means fully opaque. This is used in [Wikipedia](https://en.wikipedia.org/wiki/Wavefront_.obj_file), and most of other softwares.

However, another definition is using Tr as an alias of d, with the same usage. So the fully opaque is d = Tr = 1. This implementation is recorded in [andrewnoske‘s wiki](http://www.andrewnoske.com/wiki/OBJ_file_format) and [another spec from Department of Scientific Computing at Florida State University (FSU)](http://web.archive.org/web/20131108082758/http://people.sc.fsu.edu/~jburkardt/data/mtl/mtl.html) (They delete it at early 2014. So this link is the cache page from web.archive.org). This definition is accepted in 3ds Max, vcglib and MeshLab.

##### Three.js version

- [x] Dev
- [x] r88
- [ ] ...

##### Browser

- [x] All of them
- [ ] Chrome
- [ ] Firefox
- [ ] Internet Explorer

##### OS

- [x] All of them
- [ ] Windows
- [ ] macOS
- [ ] Linux
- [ ] Android
- [ ] iOS

##### Hardware Requirements (graphics card, VR Device, ...)

- [x] All of them

### PR DESCRIPTION
**Pull request by wtlyu**
Fixes #12861

### CODE REVIEW

**Comment by eastpiger** [comment]
Issue here #12861

**Comment by mrdoob** [comment]
I don't know...

**Comment by eastpiger** [comment]
@mrdoob  I have re-committed this modification. This modification won't change the default behavior of MTLLoader. Only add a new option to give us capability to switch the parsing of keyword Ts (Ts = 1 - d or Ts = d).

Related documents is also updated.

File: examples/js/loaders/MTLLoader.js [hunk 1]
```diff
@@ -464,7 +464,10 @@ THREE.MTLLoader.MaterialCreator.prototype = {
 
 					if ( n > 0 ) {
 
-						params.opacity = 1 - n;
+						if ( this.options && this.options.equalizeTransparency )
```
Comment (Mugen87) [inline examples/js/loaders/MTLLoader.js]: Just a code style issue. Please write it like this
```javascript
if ( this.options && this.options.equalizeTransparency ) {

   params.opacity = n;

} else {

   params.opacity = 1 - n;

}

**Review by Mugen87 — CHANGES_REQUESTED** [review]

File: examples/js/loaders/MTLLoader.js (hunk 1 above)
Comment (eastpiger) [reply examples/js/loaders/MTLLoader.js]: ok, I will update soon

**Comment by WestLangley** [comment]
Did you test this code?

**Comment by WestLangley** [comment]
Maybe

    options.TrOneIsOpaque = true; // default is false

**Comment by eastpiger** [comment]
@Mugen87 fixed

@WestLangley 
I have no idea about name of the option, TrOneIsOpaque seems better.
And I have tested this code, it works as my expected. (However I can not upload the object file I use, because of the copyrights).

mtlloader

File: examples/js/loaders/MTLLoader.js [hunk 2]
```diff
@@ -464,7 +464,16 @@ THREE.MTLLoader.MaterialCreator.prototype = {
 
 					if ( n > 0 ) {
 
-						params.opacity = 1 - n;
+						if ( this.options && this.options.TrOneIsOpaque ) {
+
+							params.opacity = n;
```
Comment (WestLangley) [inline examples/js/loaders/MTLLoader.js]: What is `params.transparent` set to when Tr is 1?

Comment (eastpiger) [reply examples/js/loaders/MTLLoader.js]: Emmmm... good comment. I didn't notice it. fix soon

Comment (eastpiger) [reply examples/js/loaders/MTLLoader.js]: Emmmm... good comment. I didn't notice it. fix soon

Comment (eastpiger) [reply examples/js/loaders/MTLLoader.js]: updated.

File: docs/examples/loaders/MTLLoader.html [hunk 3]
```diff
@@ -79,12 +79,13 @@ <h3>[method:null setMaterialOptions]( [page:Object options] )</h3>
 				<li>wrap: What type of wrapping to apply for textures. THREE.RepeatWrapping (default), THREE.ClampToEdgeWrapping, THREE.MirroredRepeatWrapping</li>
 				<li>normalizeRGB: RGBs need to be normalized to 0-1 from 0-255. Default: false, assumed to be already normalized</li>
 				<li>ignoreZeroRGBs: Ignore values of RGBs (Ka,Kd,Ks) that are all 0's. Default: false</li>
+				<li>TrOneIsOpaque: Use values 1 of Tr field for fully opaque. This option is useful for obj exported from 3ds MAX, vcglib or meshlab. Default: false</li>
```
Comment (mrdoob) [inline docs/examples/loaders/MTLLoader.html]: How about calling it `invertTr` or `invertTrProperty`?

**Comment by mrdoob** [comment]
```js
if ( this.options && this.options.invertTr ) n = 1 - n;

if ( n < 1 ) {

	params.opacity = n;
	params.transparent = true;

}
```

**Comment by Mugen87** [comment]
@eastpiger Any news about this PR?

**Comment by eastpiger** [comment]
Yep.

**Comment by mrdoob** [comment]
Thanks!

**Comment by RSpace** [comment]
Merging this PR has made any obj model material with `Tr 0.0000` specified in the mtl file completely transparent, where before it would be completely opaque. It's breaking quite a lot stuff out there.

The problem looks fairly obvious to me. It was fine to add

```
if ( this.options && this.options.invertTrProperty ) n = 1 - n;
```

but

```
					if ( n > 0 ) {

						params.opacity = 1 - n;
						params.transparent = true;

					}
```

should never have been changed to

```
					if ( n < 1 ) {

						params.opacity = n;
						params.transparent = true;

					}
```

which completely reverse the logic and makes `invertTrProperty ` work reverse of what it should have.

I'd be happy to submit a PR to fix this, unless I'm missing something?

**Comment by mrdoob** [comment]
Yes please. A PR would be great 👌

**Comment by WestLangley** [comment]
>which completely reverse the logic and makes invertTrProperty work reverse of what it should have.

https://github.com/mrdoob/three.js/issues/12861#issue-281255061 implies "what it should have" is not well-defined. (Are the docs correct?)

Consequently, `this.options.invertTrProperty` is not well-defined, either.

**Comment by RSpace** [comment]
@WestLangley Agreed. I have no need for `this.options.invertTrProperty` personally, and as such would be just a happy to submit a PR that simply reverts this PR and gets rid of `this.options.invertTrProperty`, as I would be to submit a PR that changes logic as described above, but still leaves the question of whether the logic behind `this.options.invertTrProperty` is correct open.

I assume this PR was merged for a good reason, but I'd prefer reverting it unless @eastpiger or someone else can chime in with a good explanation of what's going on here?

**Comment by WestLangley** [comment]
As implemented, `Tr` and `d` are, by default, interpreted the same: 0 is transparent, and 1 is opaque. However, as written, the documentation is not correct.

**Comment by eastpiger** [comment]
Hi all. I'm glad to make more explanation about this problem.

as I described in issue #12861, the definition of Tr option in .MTL file is confusing since it actually hasn't never been defined officially. There is a big controversy about it between academic and engineering application. I tried to ask Dr. John Burkardt, which holds the earliest webpage I can find given another different definition, and he didn't give a clear reason and answer about that. Since these two conflicting definition is both widely used in some big projects, (for example, d = Kr = 1 for opaque material is used in 3DS MAX and vcglib, which built meshLab, a good cloud point processing tool project), I give this PR to three.js hoping given a nondiscriminatory treat to them.

On the other hand. I do totally agree that this change is not supposed to break the back compatibility. I feel very very sorry to bother your work these days. Because that we had some discussion about how to introduce this feature properly and I made reset-and-recommit three times before it was merged. I used the suggestion given by mrdoob in the final version. Unfortunately, because of my carelessness, I didn't re-check the code here and it do make a reverse effect. Yes, @RSpace , you are right, that code treats the Tr option in the inverse way.

I will open a new PR fixing this soon. 

and I do want to express my apologize of my careless, and I'm truly very very sorry to @mrdoob @WestLangley @RSpace as well as anyone who was confused by my mistake.

**Comment by RSpace** [comment]
@eastpiger Thanks for the clarification. We all make mistakes, and I'm thankful to everyone who contributes to three.js, making it the fantastic library it is today. Your PR https://github.com/mrdoob/three.js/pull/13685 fixes the issue for me :)

**Comment by mrdoob** [comment]
> and I do want to express my apologize of my careless, and I'm truly very very sorry to @mrdoob @WestLangley @RSpace as well as anyone who was confused by my mistake.

No worries!! 🙌

**Pull request merged**